src/Native/Runtime/unix/PalRedhawkUnix.cpp

```diff
@@ -506,6 +509,10 @@ extern "C" void PalAttachThread(void* thread)
 extern "C" bool PalDetachThread(void* thread)
 {
     UNREFERENCED_PARAMETER(thread);
+    if (g_threadDetachCallback != nullptr)
```

Comment:
Join is supposed to be interruptible with Thread.Interrupt(), and threads wait on an event whose signal state is in managed memory. It may be possible to use a native event and consider that at any given time a thread may wait on the managed event or the native event, but other managed stuff needs to happen on thread exit as well. Mutexes owned by the exiting thread need to be abandoned, and that state is all in managed memory.

Some thoughts:
- It looks like __cxa_thread_atexit is supposed to call registered callbacks in reverse order. Can we guarantee that the call from the MRT PAL happens before a call from the CoreLib PAL?
  - If HAVE_THREAD_LOCAL is defined, thread_local is used instead of the above call. In this case I guess it would not be possible to guarantee the above. Is it possible instead to register a callback from CoreLib PAL to the MRT PAL, perhaps through managed code, upon attaching a thread to the runtime? That callback could then be raised by PalDetachThread before detaching the thread, and then I guess it should be ok to run managed code there.
- If the above is not possible and there are no other alternatives, could have a background thread that waits on a global native event in CoreLib PAL. Before thread exit, a native thread object would be queued and the event signaled. The background thread would run some managed code that would find the corresponding managed thread object and run some cleanup code, such as signaling the join event and abandoning mutexes.
- The last option if nothing else works may be to move all of the wait subsystem to the CoreLib PAL. We would lose all of the perf benefit of it being in managed code.